zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -634,48 +654,51 @@ fn create_proposed_transaction<DbT, ParamsT, InputsErrT, FeeRuleT, N>(
-    proposal_step: &proposal::Step<N>,
+    prior_step_results: &[(&Step<N>, BuildResult)],
+    proposal_step: &Step<N>,
+    #[cfg(feature = "transparent-inputs")] unused_transparent_outputs: &mut HashMap<
+        StepOutput,
+        (
+            TransparentAddress,
+            Option<TransparentAddressMetadata>,
+            OutPoint,
+        ),
+    >,
 ) -> Result<BuildResult, ErrorT<DbT, InputsErrT, FeeRuleT>>
 where
     DbT: WalletWrite + WalletCommitmentTrees,
     ParamsT: consensus::Parameters + Clone,
     FeeRuleT: FeeRule,
 {
-    // TODO: Spending shielded outputs of prior multi-step transaction steps is not yet
-    // supported. Maybe support this at some point? Doing so would require a higher-level
-    // approach in the wallet that waits for transactions with shielded outputs to be
-    // mined and only then attempts to perform the next step.
-    for s_ref in proposal_step.prior_step_inputs() {
-        prior_step_results.get(s_ref.step_index()).map_or_else(
-            || {
-                // Return an error in case the step index doesn't match up with a step
-                Err(Error::Proposal(ProposalError::ReferenceError(*s_ref)))
-            },
-            |step| match s_ref.output_index() {
-                proposal::StepOutputIndex::Payment(i) => {
-                    let prior_pool = step
-                        .0
-                        .payment_pools()
-                        .get(&i)
-                        .ok_or(Error::Proposal(ProposalError::ReferenceError(*s_ref)))?;
-
-                    if matches!(prior_pool, PoolType::Shielded(_)) {
-                        Err(Error::ProposalNotSupported)
-                    } else {
-                        Ok(())
-                    }
-                }
-                proposal::StepOutputIndex::Change(_) => {
-                    // Only shielded change is supported by zcash_client_backend, so multi-step
-                    // transactions cannot yet spend prior transactions' change outputs.
-                    Err(Error::ProposalNotSupported)
-                }
-            },
-        )?;
+    #[cfg(feature = "transparent-inputs")]
+    let step_index = prior_step_results.len();
+
+    // Spending shielded outputs of prior multi-step transaction steps (either payments or change)
+    // is not supported.
+    //
+    // TODO: Maybe support this at some point? Doing so would require a higher-level approach in
+    // the wallet that waits for transactions with shielded outputs to be mined and only then
+    // attempts to perform the next step.
+    for input_ref in proposal_step.prior_step_inputs() {
+        let prior_pool = prior_step_results
+            .get(input_ref.step_index())
+            .and_then(|(prior_step, _)| match input_ref.output_index() {
+                StepOutputIndex::Payment(i) => prior_step.payment_pools().get(&i).cloned(),
+                StepOutputIndex::Change(i) => prior_step
```

Comment:
In reading over this again, I'm having a hard time figuring out why it was that we distinguished payments and change in prior step outputs. It's all just prior step outputs, ultimately, but I think that allowing prior step payment outputs to be used here adds unnecessary complexity.

Reply:
It does, and it's a pain that they require different handling. It would be too disruptive to refactor that in this PR. In any case this code can go away once I've rebased #1441.